### ISSUE TITLE
metajam enhancements

### ISSUE THREAD

**Issue by clnsmth** [IMCR-Hackathon/ggplotgui#34]
A list of enhancements to the `metajam` project. Care must be taken to accord with maintainer preferences. [We will use this fork to develop from](https://github.com/clnsmth/metajam).

**Comment by clnsmth** [comment]
__Ignore header lines:__ Some tables have more than 1 header line and can't be read without knowing the exact number. The number of header lines may be extracted from table metadata (if present) and used to parameterize `read_csv()` and friends.

@atn38, do you have a solution for this?

**Comment by clnsmth** [comment]
__Shorten file names:__ @atn38 pointed out that paths longer than 260 characters can't be easily handled by the Windows OS (see #40). I've made one pass at shortening `metajam` directory and file names with b6512aa, but other measures may be necessary. Is there an alternative output file organization scheme that would accord with `metajam` maintainers? I'll open an issue with `metajam` maintainers.

**Comment by clnsmth** [comment]
__Download of LTER and EDI data packages is not supported:__ [This issue has been addressed here](https://github.com/IMCR-Hackathon/Hackathon-Central-2019/issues/13#issue-454988317), and needs to be submitted as a pull request to the `metajam` project.

**Comment by atn38** [comment]
**Convert missing codes (if specified in metadata) to NAs** 
Especially in case of numeric codes (e.g. -9999) in numeric columns, leaving these in skews plots and summaries significantly. `metajam` already reads in missing codes if available from attribute metadata. In `make_entity_report` I wrote a [small function here](https://github.com/IMCR-Hackathon/ggplotgui/blob/dev_summary_report/R/use_missing_code.R) to attempt to substitute missing codes with NAs. 

It'd be best if this process happens in `data_package_read` after the call to `metajam::read_d1_files`. I don't think an edit either directly in the function `read_d1_files` or to its call in `data_package_read` would work. This is because missing codes, if specified, are specific to each column; as far as I know the `read_csv/fread/etc` family of functions do not support column-specific NA strings. 

**Re: numHeaderLines**
`read_d1_files` uses `read_csv` as the default read function. It might be difficult to account for user-supplied read functions, but in the default case we can modify `metajam::download_d1_files` to read in `numHeaderLines` then supply that value to the `read_csv` function call in `read_d1_files`. I'll make a pass at this on your `metajam` fork.

**Comment by clnsmth** [comment]
Thanks for these ideas @atn38 

Regarding __Convert missing codes (if specified in metadata) to NAs__, this could be added to `metajam::read_d1_files()` [at line 60](https://github.com/clnsmth/metajam/blob/master/R/read_d1_files.R) (because at this point in the function both the data and metadata have been read, the missing value codes can be determined and replaced column wise), or we could added to `data_package_read()` as you recommend. I suggest merging your change into our project master branch, and that we at least ask `metajam` maintainers if they'd like this contribution. IMO it's a valuable contribution others will benefit from.

I agree with your recommendations for __numHeaderLines__. Another valuable enhancement.

**Comment by atn38** [comment]
@clnsmth Good point about line 60 of `read_d1_files`. It solves the "we don't know the read function" issue. Shall we open a new issue for missing codes on the main `metajam` branch? I recall there's one open for numHeaderLines already.

**Comment by clnsmth** [comment]
@atn38, IMO we should finish development of our app (at least the beta-release) before proposing changes to `metajam` incase we identify other changes that would modify the ones we already have.

**Comment by clnsmth** [comment]
**Is download_d1_data_pkg() altering file extensions?:** The file `ntl129_3_v6.csv` of `doi:10.6073/pasta/9bced2f6ff81aa30f0f573766c0a410b` is downloaded as `ntl129_3_v6.data`.

**Comment by clnsmth** [comment]
__Attempts at reading non-tabular shouldn't be made:__ `metajam` tries parsing .R files as tabular data (e.g. https://github.com/IMCR-Hackathon/datapie/issues/36#issuecomment-523551169).

**Comment by atn38** [comment]
Summary of progress on the various standing `metajam`-related enhancements and fixes so far: 

- Substitute missing codes as specified in EML with NAs: implemented in `datapie::use_missing_code()`, which is called in `data_package_read()`.
- Use number of header lines as specified in EML as number of lines to skip when parsing CSV: implemented in @clnsmth's fork of `metajam`.
- Sometimes CSV files get renamed as `.data` files which lead to parsing issues: after some investigation, this issue is likely due to how DataOne stores large CSV files. E.g. the file obtainable from D1 is a different format than the one from EDI/LTER portal. Conclusion: this is bigger than our scope and needs to be addressed by the D1 folks.

**Comment by atn38** [comment]
Update on `.data` files issue: commit a2e24fc5ed69aab097967a1614a930c7a9a0096f solves this issue via using `data.table::fread` on ".data" files where they actually are CSV files.

At current stage if I'm not mistaken, all identified and proposed enhancements have been implemented either at clnsmth/metajam or here in datapie. Only thing is to test them, of course.

**Comment by clnsmth** [comment]
Pull requests for all currently implemented changes at clnsmth/metajam have been submitted to NCEAS/metajam except the skip lines enhancements [1](https://github.com/clnsmth/metajam/commit/60fcf2594f3699e77547a477dac9d52bcec65c6c#diff-e1e1d3d40573127e9ee0480caf1283d6) and [2](https://github.com/clnsmth/metajam/commit/c8f87d95062f2cc9098e20fb32df2047b4515546#diff-e1e1d3d40573127e9ee0480caf1283d6), which will be submitted after the current batch are processed.
